src/back-end/lib/db.ts

```diff
@@ -654,3 +656,434 @@ export async function hasFilePermission(connection: Connection, session: Session
+    if (!result) {
+      throw new Error(); // to be caught by calling function
+    }
+    return result;
+  }));
+  const addenda = await Promise.all(addendaIds.map(async id => {
+    const result = getValidValue(await readOneCWUOpportunityAddendum(connection, id), null);
+    if (!result) {
+      throw new Error(); // to be caught by calling function
+    }
+    return result;
+  }));
+
+  return {
+    ...raw,
+    createdBy: createdBy || undefined,
+    updatedBy: updatedBy || undefined,
+    attachments,
+    addenda
+  };
+}
+
+export const readManyCWUOpportunities = tryDb<[Session], CWUOpportunitySlim[]>(async (connection, session) => {
+  // Retrieve the opportunity and most recent opportunity status
+
+  let query = connection<RawCWUOpportunitySlim>('cwuOpportunities as opp')
+    // Join on latest CWU status
+    .join<RawCWUOpportunitySlim>('cwuOpportunityStatuses as stat', function() {
+      this
+        .on('opp.id', '=', 'stat.opportunity')
+        .andOn('stat.createdAt', '=',
+          connection.raw('(select max("createdAt") from "cwuOpportunityStatuses" as stat2 where \
+            stat2.opportunity = opp.id)'));
+    })
+    // Join on latest CWU version
+    .join<RawCWUOpportunitySlim>('cwuOpportunityVersions as version', function() {
+      this
+        .on('opp.id', '=', 'version.opportunity')
+        .andOn('version.createdAt', '=',
+          connection.raw('(select max("createdAt") from "cwuOpportunityVersions" as version2 where \
+            version2.opportunity = opp.id)'));
+    })
+    // Select fields for 'slim' opportunity
+    .select<RawCWUOpportunitySlim[]>(
+      'opp.id',
+      'version.title',
+      'opp.createdBy',
+      'opp.createdAt',
+      'version.createdAt as updatedAt',
+      'version.createdBy as updatedBy',
+      'stat.status'
+    );
+
+  if (!session.user || session.user.type === UserType.Vendor) {
+    // Anonymous users and vendors can only see public opportunities
+    query = query
+      .whereIn('stat.status', publicOpportunityStatuses);
+  } else if (session.user.type === UserType.Government) {
+    // Gov users should only see private opportunities they own, and public opportunities
+    query = query
+      .whereIn('stat.status', publicOpportunityStatuses)
+      .orWhere(function() {
+        this
+          .whereIn('stat.status', privateOpportunitiesStatuses)
+          .andWhere({ 'opp.createdBy': session.user?.id });
+      });
+  } else {
+    // Admin users can see both private and public opportunities
+    query = query
+      .whereIn('stat.status', [...publicOpportunityStatuses, ...privateOpportunitiesStatuses]);
+  }
+
+  const results = await query;
+  // Remove createdBy/updatedBy for non-admin or non-author
+  results.map(result => {
+    if (session?.user?.type !== UserType.Admin &&
+      session?.user?.id !== result.createdBy &&
+      session?.user?.id !== result.updatedBy) {
+        delete result.createdBy;
+        delete result.updatedBy;
+    }
```

Comment:
If the user isn't signed in, the `createdBy` and `updatedBy` properties will not be deleted. They should be deleted for anonymous users.